### ISSUE TITLE
Something having a ripple effect prevents tooltip

### ISSUE THREAD

**Issue by OvermindDL1** [debois/elm-mdl#118]
Buttons, icons, it seems that anything that has a Ripple effect applied to it then tooltips will not work, rather breaking...

**Comment by OvermindDL1** [comment]
Actually this may be wrong, I cannot seem to get tooltips working anywhere, even a div...
```elm
Grid.cell
                [ Grid.size Grid.All 6
                , Options.cs "mdl-typography--caption mdl-typography--text-right"
                , Tooltip.attach Mdl (pidx ++ [ 0 ])
                ]
                [ text s_diff
                , Tooltip.render Mdl
                    (pidx ++ [ 0 ])
                    mdl
                    []
                    [ text (toString inserted_at) ]
                ]
```

**Comment by OvermindDL1** [comment]
This is not rendering either:
```elm
Button.render Mdl
            pidx
            mdl
            [ Button.icon
              -- , Button.ripple
            , button_onClick (UserlistOption msgType)
            ]
            [ Icon.view iconname [ Tooltip.attach Mdl pidx ]
            , Tooltip.render Mdl pidx mdl [] [ text name ]
            ]
```
And yet I see the tooltip div in the DOM, it just never seems to appear.

**Comment by OvermindDL1** [comment]
Continuing with the the div with text inside it example above, I do see that mouseenter/mouseleave events are registered to that div in the DOM, yet it does not seem to be working...  Continuing research...

**Comment by vipentti** [comment]
I believe currently (also in upstream) the call to `Tooltip.render` has to be a sibling element to the target.

For example:

```elm 
[ Button.render Mdl [0] model.mdl
    [ Button.raised
    , Tooltip.attach Mdl [3]
    ]
    [ text "Click me!"]
, Tooltip.render Mdl [3] model.mdl
    [ Tooltip.top
    , Tooltip.large
    ]
    [ text "Tooltips also work with material components" ]
]
```

*NOTE* that the `Tooltip.render` is the next element after the target.

**Comment by OvermindDL1** [comment]
Ahh, just as you posted that I found that in the code, altered my tests and they worked.  Thanks much, that was odd to track down.  Any chance on a compilation error or something if it is done wrong like how I did it?

**Comment by vipentti** [comment]
I think the documentation regarding `Tooltips` could be improved to make note if this rather important part. 

Compilation error would be nice but currently I don't think there is a way of doing it.

**Comment by OvermindDL1** [comment]
Maybe instead of Tooltips being as they are, maybe do something like this instead:
```elm
Tooltip.tooltip
  (text "This is the element that the tooltip will popup from, could be anything" )
  (text "And I am the tooltip that will popup, I could also be anything")
```
Would work better I'd think?  Would make it impossible to mis-use it (unless using the old Tooltip.attach and such parts if they are left in).

**Comment by OvermindDL1** [comment]
Also a question, why is it necessary to pass a `[0]` or whatever to both the attach and render if the attach does not look-up the render based on that value so the tooltip could be anywhere?

**Comment by OvermindDL1** [comment]
I have some icon buttons, when they have a ripple attached, no tooltip, when ripple is removed, tooltip works.  Their code:
```elm
          Button.render Mdl
            pidx
            mdl
            [ Button.icon
            , Button.ripple
            , button_onClick (MessageOption msgType)
            , Options.cs ("messenger-input-options-option-btn messenger-input-options-option-" ++ classname)
            ]
            [ Icon.view iconname [ Tooltip.attach Mdl (pidx++[0]) ]
            , Tooltip.render Mdl (pidx++[0]) mdl [] [ text name ]
            ]
```
If I remove the `, Button.ripple` part then the tooltip works, else it does not work.  Confirmed over repeated tests.

**Comment by vipentti** [comment]
Can confirm this as well. Might be related to #119 and Click and mouse event propagation.

**Comment by vipentti** [comment]
As a workaround what seems to work is having the tooltip attached to the icon button it self, not the inner icon.

e.g. 
```elm
    [ Button.render Mdl
        [99]
        model.mdl
        [ Button.icon
        , Button.ripple
        , Tooltip.attach Mdl [99]
        ]
        [ Icon.view "add" []
        ]

    , Tooltip.render Mdl [99] model.mdl
        []
        [ text "This is an add icon" ]
    ]

```

**Comment by OvermindDL1** [comment]
Hmm, would not fit well in the structure of the call tree but can hack it in for now.  The code I gave above is a collapsed call of my functions to give something easily testable.  Will do for now though.  :-)

**Comment by OvermindDL1** [comment]
Ahh, and sadly it will not work on my Menu one:
```elm
  Grid.cell
    [ Grid.size Grid.All 1
    , Options.cs ("messenger-input-options-menu-container messenger-input-options-menu-"++classname++"-container")
    ]
    [ Menu.render Mdl pidx mdl
        [ Menu.topRight
        , Menu.icon iconname
        --, Menu.ripple
        , Options.cs ("messenger-input-options-menu messenger-input-options-menu-"++classname)
        , Tooltip.attach Mdl (pidx++[0])
        ]
        items
    , Tooltip.render Mdl (pidx++[0]) mdl [] [ text name ]
    ]
```
As the parent of the menu is a Grid.cell, and I've not clue how to add the tooltip as a sibling to a Cell as it itself is not a Cell.  Probably have to create an intermediary `div` with a height/width at 100%...

**Comment by vipentti** [comment]
I'm currently experimenting and trying to find a way to fix this issue. It might require some horrible backs until the new Elm DOM APIs become available

**Comment by OvermindDL1** [comment]
> I'm currently experimenting and trying to find a way to fix this issue. It might require some horrible backs until the new Elm DOM APIs become available

Yeah I keep running into issue after issue of the html design.  My ports have exploded in number (thankfully all but one are 'pure', as in they take a value and/or return a value without mutating or 'doing' anything).

**Comment by vipentti** [comment]
This issue is also present on upstream MDL. 

[JSFiddle here](https://jsfiddle.net/f5qwqebg/)

**Comment by OvermindDL1** [comment]
Fascinating, wonder how that went missed...

**Comment by aforemny** [comment]
This seems to be resolved in new-v8 (probably because of f2c6628f63a6ed28113ef29be597b8cd13a3dc72). Thanks!

PS. I tested Tooltips on Buttons with Ripple. Tooltips on Menus seem to be a whole other story. :)
